### DISCUSSION TITLE
Ban on using spells in the instant battle

### DISCUSSION THREAD

**Discussion by oleg-derevenetz** [ihhub/fheroes2#2905] [Ideas]
Hi all and @idshibanov 

What do you think about the ban on the use of spells by the player's hero in the instant battle (on the world map) when instant battle is enabled in the settings? Similar behavior exists in HoMM3 HotA. The idea is that instant battle is intended to save your time, and the loss of mana is undesirable. However, if you press "auto" button during the battle, use of spells may be allowed.

**Comment by Branikolog** [comment]
Hello, @oleg-derevenetz!
Nice idea!
On one hand, this solution prevents mana from being drained by numerous battles in late game.
But on the other hand - in battles with poor armies this can affect on a chance of winning, because in battles with such circumstances first round can be determinative...
Can we just adjust the chance of using spells in a first round according to the enemy army's threat? (in a case enemy is weak, do not use spell the first round)

**Reply by oleg-derevenetz** [reply]
@Branikolog There is a PR #2896 on the approach that allows replaying a battle manually if initial auto outcome is negative. So you will be able to replay the battle if auto outcome without spells is not good enough.

**Reply by LeHerosInconnu** [reply]
Hello @oleg-derevenetz and @Branikolog,

I think the auto-combat AI should be smart enough to decide not to use magic if it does not result in the loss of creatures.
With the setting of automatic spells, magic should not necessarily be used, it should just be a possibility and not an obligation.

And the CANCEL button should be accessible in any case.

**Reply by oleg-derevenetz** [reply]
@LeHerosInconnu Unfortunately, the battle outcome window contains info about army losses, but not about mana spent. So when you will be able to view mana points of your hero, it may be already too late to press CANCEL :)

**Reply by Branikolog** [reply]
Agreed, but we need to make auto battles maximum efficient. What is the case of using auto battle and have guaranteed lost in a battle, where single spellcasting is required? I mean in a mode, where spellcasting is on. (ex: 1 vs 1 goblin)

**Reply by LeHerosInconnu** [reply]
> @LeHerosInconnu Unfortunately, the battle outcome window contains info about army losses, but not about mana spent. So when you will be able to view mana points of your hero, it may be already too late to press CANCEL :)

The number of spell points used during the combat can be simply added in the combat report window. :)

> Agreed, but we need to make auto battles maximum efficient. What is the case of using auto battle and have guaranteed lost in a battle, where single spellcasting is required? I mean in a mode, where spellcasting is on. (ex: 1 vs 1 goblin)

It depends on the attack and defense of the two heroes and which Goblin will attack first.
All this has to be determined and if it is necessary to use magic in order not to lose the combat, then a spell has to be cast.

**Reply by oleg-derevenetz** [reply]
> The number of spell points used during the combat can be simply added in the combat report window. :)

That's why I mention @idshibanov here. As far as I understand, he is mainly engaged in AI and auto battle development, so it will be interesting to hear his thoughts.

**Reply by LeHerosInconnu** [reply]
> > The number of spell points used during the combat can be simply added in the combat report window. :)
> 
> That's why I mention @idshibanov here. As far as I understand, he is mainly engaged in AI and auto battle development, so it will be interesting to hear his thoughts.

Yes, @idshibanov is the master of AI.
And he will make sure that everything is working properly. :D

**Reply by ihhub** [reply]
A replay of instant battle shouldn't be auto. The button in #2896 is designed to replay the battle in a manual way. We mustn't allow users to infinitely replay the same battle to get the best result. Please correct me here.

I don't really understand the meaning of first battle: is it first try of instant battle? If yes then how many attempts do we allow to have?

**Reply by LeHerosInconnu** [reply]
The player will replay the combat if the result of the automatic combat (adventure map) does not suit him.
At the end of the replayed combat, he will not be able to replay the combat again.

However, during the replay, the player should be able to activate the automatic combat mode as desired (combat screen).
For example, he can start some actions manually, then switch to automatic mode for a few turns and then switch back to manual mode to finish the combat.
This is how I see it, anyway.

**Reply by oleg-derevenetz** [reply]
@ihhub The idea is to ban spell usage by player's hero AI on instant battle (like in HotA) to avoid unnecessary mana drain on trivial battles, but allow usage of spells if player press CANCEL button in battle results dialog and then press AUTO button in the bottom left corner of battle screen.

**Reply by Branikolog** [reply]
We have a slight mess with "auto" and "instant" battles here.)

**Reply by Branikolog** [reply]
Oh, actually, I thought, that @oleg-derevenetz proposed to ban spells in the first round of battle in instant mode. So AI would use spells only for second round.... o_O My fault... But Idea is interesting anyway! )

**Comment by idshibanov** [comment]
Hi everyone!

I'm not sure why need this logic because we already have an option to disable spell casting in auto combat. Also I don't like "hidden" behaviors and "first auto combat" is very vague. Is it a first one during the gaming session? First one when you enable the setting? Why only first? Intent should be clear - if you want to save mana just tell it instead of flipping settings back and forth.

![image](https://user-images.githubusercontent.com/1373638/110211131-2c002080-7e63-11eb-99cd-fcca4d81d13e.png)

In addition AI has a check internally to conserve hero's mana if battle can be easily won, so it shouldn't spam spells if you meet 1 goblin.

**Reply by LeHerosInconnu** [reply]
Yes, it should be simple and clear to the player.
And there is also a setting for casting spells for the automatic combat (combat screen).

In fheroes2:

![Spell casting combat screen fh2 01](https://user-images.githubusercontent.com/43583994/110211404-bf9f1480-7e96-11eb-8c56-cb19f7dc14c5.png)

**Reply by oleg-derevenetz** [reply]
Hi @idshibanov

I probably didn't put it quite correctly. By "first auto battle" I mean instant battle. That setting, if I understand correctly, disables spells during any auto battle, including both instant battles and auto battle when player press AUTO button in the bottom left corner of the battle screen. The problem is that currently player didn't know for sure was the mana spent during the instant battle or not.

**Reply by idshibanov** [reply]
Alright, now I understand. Never thought of such a use case myself because if I would press "Cancel" I will do everything manually. Yes, right now auto spell casting controls both because logically "instant" and "auto" is the same thing. It fully plays out your battle in the background just won't load the interface.

Regarding mana: it's always challenging to add something new to an old interface because pixel space is very limited. Outputting the number is easy but making it look good will take some brainstorming. Even such trivial task as adding "Cancel" button took some time because existing "Okay" button was already drawn on top of dialog background which we had to overwrite.

**Reply by oleg-derevenetz** [reply]
@idshibanov I'm aware of troubles with adjusting original UI, that's why I suggest to disable spells in instant battle entirely. But this, of course, is debatable.

**Reply by LeHerosInconnu** [reply]
A simple line indicating the number of spell points spent and the number of spell points remaining in relation to the hero's total spell points would allow the player to make an informed decision:

![Battle report with spell points fh2 01](https://user-images.githubusercontent.com/43583994/110214940-45c35700-7ea7-11eb-807c-0edc8afc74c4.png)

**Comment by oleg-derevenetz** [comment]
Paraphrased header & initial message to avoid confusion between auto battle and instant battle.
